zerver/actions/message_send.py

```diff
@@ -205,6 +205,7 @@ def get_recipient_info(
         # of this function for different message types.
         assert stream_topic is not None
         user_ids_muting_topic = stream_topic.user_ids_with_visibility_policy(UserTopic.MUTED)
+        user_ids_unmuting_topic = stream_topic.user_ids_with_visibility_policy(UserTopic.UNMUTED)
```

Comment:
I think we discussed this on the previous PR, but we should do a preparatory refactor to make a `StreamTopicTarget` just query all user IDs with a visibility policy once. Maybe it'd work to just do that query when it is initialized, and have it put the objects returned into a dict of sets that can be accessed with this API, but I'm not particular about the details other than that we avoid duplicate database queries here.

Reply:
I tried this approach.

But the problem is, the dictionary doesn't contains updated data. 
Like, In tests we continuosly update the `visibility_policy`, but dictionary which was created during `StreamTopicTarget` initialization has older data.

But, I think we should probably find a way to optimize it. Becase, in future, with the increase in `visibility_policy` choices. query count will keep increasing.

Reply:
> The StreamTopicTarget structure is relevant for only a single message-send request,

Then we should configure separate smaller tests out of [test_stream_recipient_info](https://github.com/zulip/zulip/pull/24236/commits/cc5a03241b356504cf33e9a3389bab5cad5a1de3#diff-338fc8375bdda13ccdd1d4a31b273f2ee9a121f0fbe7bffe820b8b3d24f43001R1859) (Especially when we are updating visibility_policy) 

It uses the same [StreamTopicTarget](https://github.com/zulip/zulip/blob/main/zerver/tests/test_users.py#L1757) object throughout the tests, resulting in [assertion error](https://github.com/zulip/zulip/pull/24236/commits/cc5a03241b356504cf33e9a3389bab5cad5a1de3#diff-338fc8375bdda13ccdd1d4a31b273f2ee9a121f0fbe7bffe820b8b3d24f43001L1870) (after we change visibility policy)

Reply:
No, what I meant was, there's no need to store the result of the DB query in `StreamTopicTarget`.
It should be suffient to just modify the DB query so that all the data can be fetched by a single query instead of 2 separate ones (as explained in my comment above, last paragraph)

Reply:
So, we are handling this in the following way:

A helper function [`get_visibility_policy_to_user_ids`](https://github.com/zulip/zulip/pull/24236/commits/33506f654c5c06c9419ae2a5185e059979bd2320#diff-d803e860e5ff2068cb51aab642015247be71eb4045444a4dcb82e86486f22a4dR156) is added which returns a dictionary of the form
```
(visibility_policy=>set(user_ids))
```

The function queries the database only once.

and finally we can just do:
```
     visibility_policy_to_user_ids = get_visibility_policy_to_user_ids(stream_topic)
     user_ids_muting_topic = visibility_policy_to_user_ids[UserTopic.MUTED]
     user_ids_unmuting_topic = visibility_policy_to_user_ids[UserTopic.UNMUTED]
```

Reply:
[Updated Approach](https://github.com/zulip/zulip/pull/24236#discussion_r1105304217) -- after #24196